tests/numerics/fv/test_upwind.py

```diff
@@ -0,0 +1,682 @@
+        assert np.allclose(deltaT, deltaT_known, rtol, atol)
+        if bc_type == "dir":
+            # retrieve extra references
+            rhs_known = self._rhs_collection[test_name]
+            assert np.allclose(rhs, rhs_known, rtol, atol)
+
+
+class TestMixedDimensionalUpwind:
+
+    """Helper function for fill up dictionaries."""
+
+    def _assign_discretization(self, mdg, disc, coupling_disc, variable):
+        term = "advection"
+        for _, data in mdg.subdomains(return_data=True):
+            data[pp.PRIMARY_VARIABLES] = {variable: {"cells": 1}}
+            data[pp.DISCRETIZATION] = {variable: {term: disc}}
+
+        for intf, data in mdg.interfaces(return_data=True):
+            sd_primary, sd_secondary = mdg.interface_to_subdomain_pair(intf)
+            data[pp.PRIMARY_VARIABLES] = {"lambda_u": {"cells": 1}}
+            data[pp.COUPLING_DISCRETIZATION] = {
+                variable: {
+                    sd_primary: (variable, term),
+                    sd_secondary: (variable, term),
+                    intf: ("lambda_u", coupling_disc),
+                }
+            }
+
+    """Helper function for adding a vector flux for faces."""
+
+    def _add_constant_darcy_flux(self, mdg, upwind, flux, a):
+        for sd, data in mdg.subdomains(return_data=True):
+            aperture = np.ones(sd.num_cells) * np.power(a, mdg.dim_max() - sd.dim)
+            data[pp.PARAMETERS]["transport"]["darcy_flux"] = upwind.darcy_flux(
+                sd, flux, aperture
+            )
+        for intf, data in mdg.interfaces(return_data=True):
+            sd_primary, _ = mdg.interface_to_subdomain_pair(intf)
+            param_primary = mdg.subdomain_data(sd_primary)[pp.PARAMETERS]
+            darcy_flux = param_primary["transport"]["darcy_flux"]
+            sign = np.zeros(sd_primary.num_faces)
+            boundary_faces = sd_primary.get_all_boundary_faces()
+            boundary_signs, _ = sd_primary.signs_and_cells_of_boundary_faces(
+                boundary_faces
+            )
+            sign[boundary_faces] = boundary_signs
+
+            sign = intf.primary_to_mortar_avg() * sign
+            darcy_flux_e = sign * (intf.primary_to_mortar_avg() * darcy_flux)
+            if pp.PARAMETERS not in data:
+                data[pp.PARAMETERS] = pp.Parameters(
+                    intf, ["transport"], [{"darcy_flux": darcy_flux_e}]
+                )
+            else:
+                data[pp.PARAMETERS]["transport"]["darcy_flux"] = darcy_flux_e
+
+    def _compose_algebraic_representation(self, mdg, upwind_obj, upwind_coupling_obj):
+        """Create discretization matrix and rhs vector from mdg, upwind and
+            upwind_coupling objects.
+
+        This method essentially mimics the functionality of the antiquated assembler class
+        """
+
+        sd_hashes = [hash(sd) for sd in mdg.subdomains()]
+        intf_hashes = [hash(intf) for intf in mdg.interfaces()]
+        hashes = sd_hashes + intf_hashes
```

Comment:
Do we need to use hashes here? Why not just lists?

Reply:
I think I see why (below). Please leave a short documentation (here and below).

Reply:
Thanks for the comment, I have improved the documentation.